### PR TITLE
Todo application

### PR DESCRIPTION
**Pull request by eugene-chr**

### CODE REVIEW

File: src/components/shared/IconButton/component.js [hunk 1]
```diff
@@ -0,0 +1,65 @@
+/* eslint-disable react/button-has-type */
+import { faSpinner } from '@fortawesome/free-solid-svg-icons';
+import { FontAwesomeIcon } from '@fortawesome/react-fontawesome';
+import classNames from 'classnames';
+import PropTypes from 'prop-types';
+import React from 'react';
+
+const IconButton = ({
```
Comment (plzen) [inline src/components/shared/IconButton/component.js]: Can't you reuse Button component and pass `<FontAwesomeIcon icon={icon} />` as children?

File: src/components/shared/Preloader/component.js [hunk 2]
```diff
@@ -0,0 +1,12 @@
+import React from 'react';
+import { faSpinner } from '@fortawesome/free-solid-svg-icons';
+import { FontAwesomeIcon } from '@fortawesome/react-fontawesome';
+
+const Preloader = () => (
+  <div className="preloader">
+    {/* <p>No items...</p> */}
```
Comment (plzen) [inline src/components/shared/Preloader/component.js]: Remove commented code

File: src/components/withAuth.js [hunk 3]
```diff
@@ -0,0 +1,39 @@
...
+const WithAuth = WrappedComponent => {
+  class WithAuthHOC extends Component {
+    render() {
+      const { user } = this.props;
+
+      return !user ? <Redirect to="/sign-up" /> : <WrappedComponent {...this.props} />;
+    }
+  }
+
+  const mapStateToProps = state => {
```
Comment (plzen) [inline src/components/withAuth.js]: Simplify this method

File: src/components/withoutAuth.js [hunk 4]
```diff
@@ -0,0 +1,39 @@
...
+const WithoutAuth = WrappedComponent => {
+  class WithoutAuthHOC extends Component {
+    render() {
+      const { user } = this.props;
+
+      return user ? <Redirect to="/projects" /> : <WrappedComponent {...this.props} />;
+    }
+  }
+
+  const mapStateToProps = state => {
```
Comment (plzen) [inline src/components/withoutAuth.js]: Simplify this method

File: src/store/logic.js [hunk 5]
```diff
@@ -0,0 +1,35 @@
+import { sessionLogic, logoutLogic } from './session/logic';
+import signInLogic from './signIn/logic';
+import signUpLogic from './signUp/logic';
+import {
+  saveProjectLogic, projectsLogic, deleteProjectLogic, updateProjectLogic,
+} from './projects/logic';
+import {
+  tasksLogic,
```
Comment (plzen) [inline src/store/logic.js]: You can import all logics like
```
import * as taskLogics from './tasks/logic'
```
and use spread operator
```
export default [
  ...taskLogics,
];
```

File: src/store/modal/reducers.js [hunk 6]
```diff
@@ -0,0 +1,28 @@
...
+    case OPEN_MODAL:
+      return {
+        ...state,
+        show: true,
+        props: action.payload.props,
+        type: action.payload.type,
+      };
+    case CLOSE_MODAL:
+      return {
+        show: false,
```
Comment (plzen) [inline src/store/modal/reducers.js]: I think you also need to remove `props` and `type` when modal closes

File: src/store/projects/logic.js [hunk 7]
```diff
@@ -0,0 +1,127 @@
...
+import { deleteProjectTasks } from '../tasks/actions';
+import { closeModal } from '../modal/actions';
+
+const projectsLogic = createLogic({
+  type: FETCH_PROJECTS,
+  latest: true,
+
+  async process({ getState, db }, dispatch, done) {
+    const { session } = getState();
+    const { uid } = session.user;
```
Comment (plzen) [inline src/store/projects/logic.js]: Use selector

File: src/store/rootReducer.js [hunk 8]
```diff
@@ -0,0 +1,26 @@
...
+const appReducer = combineReducers({
+  session,
+  projects,
+  tasks,
+  notifications,
+  modal,
+});
+
+const rootReducer = (state, action) => {
+  if (action.type === LOGOUT_USER) {
```
Comment (plzen) [inline src/store/rootReducer.js]: Why do you need this here? You already have handling this action in session reducer

File: src/App.js [hunk 9]
```diff
@@ -0,0 +1,71 @@
...
+import Projects from './components/Projects';
+import SignIn from './components/SignIn';
+import SignUp from './components/SignUp';
+import { isUserLoggedIn } from './store/session/actions';
+import Notifications from './components/Notifications';
+import logo from './assets/images/logo.svg';
+import ModalRoot from './components/ModalRoot';
+import { modalState } from './store/modal/selectors';
+
+class App extends Component {
```
[the remark this answers is not in the document]
Comment (eugene-chr) [reply src/App.js]: done

File: package.json [hunk 10]
```diff
@@ -0,0 +1,62 @@
...
+  "version": "0.1.0",
+  "private": true,
+  "dependencies": {
+    "@fortawesome/fontawesome-svg-core": "^1.2.22",
+    "@fortawesome/free-solid-svg-icons": "^5.10.2",
+    "@fortawesome/react-fontawesome": "^0.1.4",
+    "bootstrap": "^4.3.1",
+    "classnames": "^2.2.6",
+    "dayjs": "^1.8.16",
+    "eslint-config-airbnb": "^18.0.1",
```
[the remark this answers is not in the document]
Comment (eugene-chr) [reply package.json]: done

File: src/components/Notifications/NotificationItem/container.js [hunk 11]
```diff
@@ -0,0 +1,56 @@
...
+    this.timer();
+  }
+
+  componentWillUnmount() {
+    const { timer } = this.state;
+
+    clearTimeout(timer);
+  }
+
+  timer = () => {
```
[the remark this answers is not in the document]
Comment (eugene-chr) [reply src/components/Notifications/NotificationItem/container.js]: done

File: src/components/firebase.js [hunk 12]
```diff
@@ -0,0 +1,15 @@
+import app from 'firebase/app';
+import 'firebase/auth';
+import 'firebase/firestore';
+
+const firebaseConfig = {
```
[the remark this answers is not in the document]
Comment (eugene-chr) [reply src/components/firebase.js]: done

File: src/store/modal/reducers.js (hunk 6 above)
Comment (eugene-chr) [reply src/store/modal/reducers.js]: done

File: src/components/Header/container.js [hunk 13]
```diff
@@ -0,0 +1,34 @@
...
+  render() {
+    return <HeaderComponent {...this.props} />;
+  }
+}
+
+const mapDispatchToProps = {
+  logoutUser,
+};
+
+const mapStateToProps = state => {
```
[the remark this answers is not in the document]
Comment (eugene-chr) [reply src/components/Header/container.js]: done

File: src/components/ModalRoot/container.js [hunk 14]
```diff
@@ -0,0 +1,20 @@
...
+
+// eslint-disable-next-line react/prefer-stateless-function
+class ModalRoot extends React.Component {
+  render() {
+    return <ModalRootComponent {...this.props} />;
+  }
+}
+
+const mapStateToProps = state => ({
+  type: state.modal.type,
```
[the remark this answers is not in the document]
Comment (eugene-chr) [reply src/components/ModalRoot/container.js]: done

File: src/components/ModalRoot/container.js [hunk 15]
```diff
@@ -0,0 +1,20 @@
...
+  modalProps: state.modal.props,
+});
+
+export default connect(
+  mapStateToProps,
+  null,
```
[the remark this answers is not in the document]
Comment (eugene-chr) [reply src/components/ModalRoot/container.js]: done

File: src/components/Notifications/container.js [hunk 16]
```diff
@@ -0,0 +1,36 @@
...
+class Notifications extends Component {
+  render() {
+    const { notifications } = this.props;
+
+    return ReactDOM.createPortal(<NotificationsComponent notifications={notifications} />, document.body);
+  }
+}
+
+const mapStateToProps = state => {
+  const notifications = getNotifications(state);
```
[the remark this answers is not in the document]
Comment (eugene-chr) [reply src/components/Notifications/container.js]: done

File: src/components/Projects/ProjectForm/component.js [hunk 17]
```diff
@@ -0,0 +1,79 @@
...
+    setFieldTouched,
+    handleReset,
+    buttons,
+    handleFocus,
+  } = props;
+
+  return (
+    <form onSubmit={handleSubmit}>
+      <div className="task__form ml-0">
+        <Input
```
[the remark this answers is not in the document]
Comment (eugene-chr) [reply src/components/Projects/ProjectForm/component.js]: done

File: src/components/Projects/ProjectItem/ProjectItemForm/component.js [hunk 18]
```diff
@@ -0,0 +1,69 @@
...
+    isSubmitting,
+    setFieldTouched,
+    touched,
+    values,
+  } = props;
+
+  return (
+    <form onSubmit={handleSubmit} className="task__form ml-26">
+      <div className="task__form ml-0">
+        <Input
```
[the remark this answers is not in the document]
Comment (eugene-chr) [reply src/components/Projects/ProjectItem/ProjectItemForm/component.js]: odne

File: src/components/Projects/ProjectItem/ProjectModalConfirmDelete/container.js [hunk 19]
```diff
@@ -0,0 +1,39 @@
...
+
+    deleteProject(projectID);
+  };
+
+  render() {
+    return <TaskModalConfirmDeleteComponent handleDeleteProject={this.handleDeleteProject} />;
+  }
+}
+
+const mapStateToProps = state => {
```
[the remark this answers is not in the document]
Comment (eugene-chr) [reply src/components/Projects/ProjectItem/ProjectModalConfirmDelete/container.js]: done

File: src/components/Projects/ProjectList/component.js [hunk 20]
```diff
@@ -0,0 +1,9 @@
+import React from 'react';
+import Loading from 'components/loading';
+import ProjectItem from '../ProjectItem';
+
+const ProjectList = ({ projects }) => projects.map(project => (
+  <ProjectItem key={project.id} projectID={project.id} projectName={project.projectName} />
+));
+
+export default Loading(ProjectList);
```
[the remark this answers is not in the document]
Comment (eugene-chr) [reply src/components/Projects/ProjectList/component.js]: done

File: src/components/Projects/container.js [hunk 21]
```diff
@@ -0,0 +1,85 @@
...
+  projects: PropTypes.arrayOf(PropTypes.shape({
+    projectName: PropTypes.string,
+    id: PropTypes.string,
+    uid: PropTypes.string,
+    tasks: PropTypes.arrayOf(PropTypes.string),
+  })),
+  projectName: PropTypes.string,
+};
+
+export default connect(
```
[the remark this answers is not in the document]
Comment (eugene-chr) [reply src/components/Projects/container.js]: done

File: src/components/Tasks/TaskItem/TaskItemForm/component.js [hunk 22]
```diff
@@ -0,0 +1,62 @@
...
+          error={errors.taskName}
+          handleBlur={setFieldTouched}
+          handleChange={handleChange}
+          handleFocus={handleFocus}
+          name="taskName"
+          placeholder="Enter task name"
+          touched={touched.taskName}
+          value={values.taskName}
+        />
+        {true && (
```
[the remark this answers is not in the document]
Comment (eugene-chr) [reply src/components/Tasks/TaskItem/TaskItemForm/component.js]: done

File: src/components/Tasks/TaskItem/TaskModalComments/TaskModalCommentsForm/container.js [hunk 23]
```diff
@@ -0,0 +1,65 @@
...
+import TaskModalCommentsFormComponent from './component';
+
+class TaskModalCommentsForm extends Component {
+  state = {
+    fileName: '',
+    thumb: null,
+  };
+
+  handleSubmit = () => {
+    console.log('kek');
```
[the remark this answers is not in the document]
Comment (eugene-chr) [reply src/components/Tasks/TaskItem/TaskModalComments/TaskModalCommentsForm/container.js]: done

File: src/components/Tasks/TaskItem/TaskModalComments/TaskModalCommentsForm/container.js [hunk 24]
```diff
@@ -0,0 +1,65 @@
...
+    const { showNotification } = this.props;
+    const { size, name, type } = fileRef;
+    const MAX_FILE_SIZE = 1e+7;
+    const TYPES = ['image/png', 'image/jpeg'];
+    this.setState({ fileName: name });
+    console.log(event.target.files[0]);
+    if (size > MAX_FILE_SIZE) {
+      showNotification('An uploaded file is too large. The size shouldn’t exceed 10 MB.', 'error');
+    }
+    console.log(some(TYPES, type));
```
[the remark this answers is not in the document]
Comment (eugene-chr) [reply src/components/Tasks/TaskItem/TaskModalComments/TaskModalCommentsForm/container.js]: done

File: src/components/Tasks/TaskItem/TaskModalConfirmDelete/container.js [hunk 25]
```diff
@@ -0,0 +1,46 @@
...
+
+    deleteTask(taskID, projectID);
+  };
+
+  render() {
+    return <TaskModalConfirmDeleteComponent handleDeleteTask={this.handleDeleteTask} />;
+  }
+}
+
+const mapStateToProps = state => {
```
[the remark this answers is not in the document]
Comment (eugene-chr) [reply src/components/Tasks/TaskItem/TaskModalConfirmDelete/container.js]: done

File: src/components/Tasks/TaskItem/TaskModalDeadline/component.js [hunk 26]
```diff
@@ -0,0 +1,19 @@
...
+import { Datepicker, Button } from 'components/shared';
+
+const TaskModalDeadlineComponent = ({ handleSetDateline, date, time }) => (
+  <Modal>
+    <ModalHeader>Deadline</ModalHeader>
+    <ModalBody>
+      <Datepicker date={date} time={time} />
+    </ModalBody>
+    <ModalFooter>
+      <Button type="submit" color="green" onClick={() => {}}>Save</Button>
```
[the remark this answers is not in the document]
Comment (eugene-chr) [reply src/components/Tasks/TaskItem/TaskModalDeadline/component.js]: done

File: src/components/Tasks/TaskItem/TaskModalDeadline/container.js [hunk 27]
```diff
@@ -0,0 +1,46 @@
...
+      <TaskModalDeadlineComponent date={date} time={time} handleSetDateline={this.handleSetDateline} />
+    );
+  }
+}
+
+const mapStateToProps = state => {
+  const { taskID } = getModalProps(state);
+  const { date, time } = getTaskDeadline(state, taskID);
+
+  // console.log(date, time);
```
[the remark this answers is not in the document]
Comment (eugene-chr) [reply src/components/Tasks/TaskItem/TaskModalDeadline/container.js]: done

File: src/components/Tasks/TaskList/component.js [hunk 28]
```diff
@@ -0,0 +1,28 @@
...
+      />
+    ))}
+  </ul>
+);
+
+TaskList.propTypes = {
+  tasks: PropTypes.array,
+};
+
+export default Loading(TaskList);
```
[the remark this answers is not in the document]
Comment (eugene-chr) [reply src/components/Tasks/TaskList/component.js]: done

File: src/components/shared/Datepicker/container.js [hunk 29]
```diff
@@ -0,0 +1,48 @@
...
+import { withFormik } from 'formik';
+import PropTypes from 'prop-types';
+import { datepickerSchema } from 'components/validationSchema';
+import DatepickerComponent from './component';
+
+class DatepickerForm extends Component {
+  onChangeDate = value => {
+    const { setFieldValue } = this.props;
+
+    // console.log('date', value);
```
[the remark this answers is not in the document]
Comment (eugene-chr) [reply src/components/shared/Datepicker/container.js]: done

File: src/components/shared/IconButton/component.js (hunk 1 above)
Comment (eugene-chr) [reply src/components/shared/IconButton/component.js]: done

File: src/components/shared/Preloader/component.js (hunk 2 above)
Comment (eugene-chr) [reply src/components/shared/Preloader/component.js]: done

File: src/components/withAuth.js (hunk 3 above)
Comment (eugene-chr) [reply src/components/withAuth.js]: done

File: src/components/withoutAuth.js (hunk 4 above)
Comment (eugene-chr) [reply src/components/withoutAuth.js]: done

File: src/store/projects/logic.js (hunk 7 above)
Comment (eugene-chr) [reply src/store/projects/logic.js]: done

**Pull request merged**